dpnp/dpnp_algo/dpnp_fill.py

```diff
@@ -0,0 +1,89 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+# THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+import dpctl.tensor as dpt
+import dpctl.utils as dpu
+import numpy as np
+from dpctl.tensor._tensor_impl import (
+    _copy_usm_ndarray_into_usm_ndarray,
+    _full_usm_ndarray,
+    _zeros_usm_ndarray,
+)
+
+import dpnp
+
+
+def dpnp_fill(arr, val):
+    arr = dpnp.get_usm_ndarray(arr)
+    exec_q = arr.sycl_queue
+
+    dpnp.check_supported_arrays_type(val, scalar_type=True, all_scalars=True)
+    # if val is an array, process it
+    if dpnp.is_supported_array_type(val):
+        val = dpnp.get_usm_ndarray(val)
+        if val.shape != ():
+            raise ValueError("`val` must be a scalar or 0D-array")
+        if dpu.get_execution_queue((exec_q, val.sycl_queue)) is None:
+            raise dpu.ExecutionPlacementError(
+                "Input arrays have incompatible queues."
+            )
+        a_val = dpt.asarray(
+            val,
+            dtype=arr.dtype,
+            usm_type=arr.usm_type,
+            sycl_queue=exec_q,
+        )
+        a_val = dpt.broadcast_to(a_val, arr.shape)
+        _manager = dpu.SequentialOrderManager[exec_q]
+        dep_evs = _manager.submitted_events
+        h_ev, c_ev = _copy_usm_ndarray_into_usm_ndarray(
+            src=a_val, dst=arr, sycl_queue=exec_q, depends=dep_evs
+        )
+        _manager.add_event_pair(h_ev, c_ev)
+        return
+    elif not dpnp.isscalar(val):
+        raise TypeError(
+            f"Expected `val` to be a 0D-array or Python scalar, got {type(val)}"
+        )
+
+    dt = arr.dtype
+    val_type = type(val)
+    if val_type in [float, complex] and dpnp.issubdtype(dt, dpnp.integer):
+        val = int(val.real)
+    elif val_type is complex and dpnp.issubdtype(dt, dpnp.floating):
+        val = val.real
+    elif val_type is int and dpnp.issubdtype(dt, dpnp.integer):
+        val = np.asarray(val, dtype=dt)[()]
+
+    _manager = dpu.SequentialOrderManager[exec_q]
+    dep_evs = _manager.submitted_events
+    # can leverage efficient memset when val is 0
+    if arr.flags["FORC"] and val == 0:
+        h_ev, zeros_ev = _zeros_usm_ndarray(arr, exec_q, depends=dep_evs)
+        _manager.add_event_pair(h_ev, zeros_ev)
+    else:
+        h_ev, fill_ev = _full_usm_ndarray(val, arr, exec_q, depends=dep_evs)
```

Comment:
below example produce an unexpected error inside the extension:
```python
a = dpnp.ones(10, dtype='i')
a.fill('0')

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[9], line 1
----> 1 a.fill('0')

File ~/code/dpnp/dpnp/dpnp_array.py:912, in dpnp_array.fill(self, value)
    909 # lazy import avoids circular imports
    910 from .dpnp_algo.dpnp_fill import dpnp_fill
--> 912 dpnp_fill(self, value)

File ~/code/dpnp/dpnp/dpnp_algo/dpnp_fill.py:92, in dpnp_fill(arr, val)
     90     _manager.add_event_pair(h_ev, zeros_ev)
     91 else:
---> 92     h_ev, fill_ev = _full_usm_ndarray(val, arr, exec_q, depends=dep_evs)
     93     _manager.add_event_pair(h_ev, fill_ev)

RuntimeError: Unable to cast Python instance of type <class 'str'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```

Reply:
`dpnp.full` produces this error, too.

```
In [1]: import dpnp as dnp

In [2]: dnp.full(10, "0", dtype="i4")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[2], line 1
----> 1 dnp.full(10, "0", dtype="i4")

File ~/repos/dpnp-fork/dpnp/dpnp_iface_arraycreation.py:2120, in full(shape, fill_value, dtype, order, like, device, usm_type, sycl_queue)
   2041 """
   2042 Return a new array of given shape and type, filled with `fill_value`.
   2043
   (...)
   2115
   2116 """
   2118 dpnp.check_limitations(like=like)
-> 2120 return dpnp_container.full(
   2121     shape,
   2122     fill_value,
   2123     dtype=dtype,
   2124     order=order,
   2125     device=device,
   2126     usm_type=usm_type,
   2127     sycl_queue=sycl_queue,
   2128 )

File ~/repos/dpnp-fork/dpnp/dpnp_container.py:230, in full(shape, fill_value, dtype, order, device, usm_type, sycl_queue)
    227     fill_value = fill_value.get_array()
    229 """Creates `dpnp_array` having a specified shape, filled with fill_value."""
--> 230 array_obj = dpt.full(
    231     shape,
    232     fill_value,
    233     dtype=dtype,
    234     order=order,
    235     usm_type=usm_type,
    236     sycl_queue=sycl_queue_normalized,
    237 )
    238 return dpnp_array._create_from_usm_ndarray(array_obj)

File ~/repos/dpctl/dpctl/tensor/_ctors.py:1128, in full(shape, fill_value, dtype, order, device, usm_type, sycl_queue)
   1126 _manager = dpctl.utils.SequentialOrderManager[sycl_queue]
   1127 # populating new allocation, no dependent events
-> 1128 hev, full_ev = ti._full_usm_ndarray(fill_value, res, sycl_queue)
   1129 _manager.add_event_pair(hev, full_ev)
   1130 return res

RuntimeError: Unable to cast Python instance of type <class 'str'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)
```
For this case to be covered, `dpnp` will need a utility for extracting Python scalar from Python string.

That can be handled here or in a separate PR.

Reply:
As for now, we state only support for scalar or an array as an input `fill_value` to `dpnp.full` function and doesn't support string input.
Thus another alternative might be to have a validation check with a proper error message in both `dpt.full` and `dpnp.ndarray.fill` functions before calling `_full_usm_ndarray` from the extension. Since I guess `RuntimeError` is not that we are mainly expecting to have in such use case.

Reply:
I have opened https://github.com/IntelPython/dpctl/pull/1878 to resolve this in `full` and `full_like`.

I will apply a similar validation check to `fill`.